Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs

```diff
@@ -111,76 +111,31 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
             }
 
             var jsonObject = JObject.Load(reader);
-            var childTypes = GetChildTypes(objectType);
-
             var jsonProperties = new HashSet<string>(GetProperties(jsonObject), StringComparer.OrdinalIgnoreCase);
-            var candidateChildTypes = new List<CandidateChildType>();
-            foreach (var childType in childTypes)
-            {
-                var childTypeProperties = childType.GetProperties().Select(t => t.Name).ToList();
-                var childTypeFields = childType.GetFields().Select(t => t.Name).ToList();
-                var childTypePropertiesAndFields = new HashSet<string>(
-                    childTypeProperties.Concat(childTypeFields),
-                    StringComparer.OrdinalIgnoreCase);
-                if (jsonProperties.All(p => childTypePropertiesAndFields.Contains(p)))
-                {
-                    var candidateChildType = new CandidateChildType
-                    {
-                        Type = childType,
-                        PropertiesAndFields = childTypePropertiesAndFields
-                    };
-                    candidateChildTypes.Add(candidateChildType);
-                }
-            }
 
-            foreach (var candidateChildType in candidateChildTypes)
-            {
-                try
-                {
-                    var deserializedObject = serializer.Deserialize(jsonObject.CreateReader(), candidateChildType.Type);
-                    if (deserializedObject != null)
-                    {
-                        candidateChildType.DeserializedObject = deserializedObject;
-                    }
-                }
-                catch (JsonException)
-                {
-                }
-            }
+            var candidateChildTypes = GetCandidateChildTypes(objectType, jsonProperties).ToList();
+
+            DeserializeCandidates(serializer, jsonObject, candidateChildTypes);
 
             candidateChildTypes = candidateChildTypes.Where(c => c.DeserializedObject != null).ToList();
```

Comment:
The trade off we didn't like was elevating that class to be external to this one because of the way Recipes are embedded making internals effectively public creating an unintended surface area.